### PR TITLE
gh-109693: Update _gil_runtime_state.locked to use pyatomic.h

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-109693

### CODE REVIEW

**Review by vstinner — APPROVED** [review]
LGTM.

@colesbury: Do you want to review pyatomic.h changes?

File: Include/cpython/pyatomic_msc.h [hunk 1]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
 #endif
 }
 
+static inline void
+_Py_atomic_store_int_release(int *obj, int value)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    *(int * volatile *)obj = value;
+#elif defined(_M_ARM64)
+    __stlr64((unsigned __int64 volatile *)obj, (int)value);
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h]: Hum, __int64 is larger than int, no?

Comment (corona10) [reply Include/cpython/pyatomic_msc.h]: > Hum, __int64 is larger than int, no?

Hmm I am checking this: https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170

Comment (corona10) [reply Include/cpython/pyatomic_msc.h]: Yeah you are right, it should be 32bit.

**Review by vstinner — COMMENTED** [review]
Windows logs many compiler warnings, see: https://github.com/python/cpython/pull/110836/files

File: Include/cpython/pyatomic_msc.h [hunk 2]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
 #endif
 }
 
+static inline void
+_Py_atomic_store_int_release(int *obj, int value)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    *(int * volatile *)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```
    *(volatile int *)obj = value;
```

File: Include/cpython/pyatomic_msc.h [hunk 3]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
+#elif defined(_M_ARM64)
+    __stlr32((unsigned __int32 volatile *)obj, (int)value);
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```
    __stlr32((unsigned __int32 volatile *)obj, (unsigned __int32)value);
```

File: Include/cpython/pyatomic_msc.h [hunk 4]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
+#endif
+}
+
+static inline int
+_Py_atomic_load_int_acquire(const int *obj)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    return *(int volatile *)obj;
+#elif defined(_M_ARM64)
+    return (int *)__ldar32((unsigned __int32 volatile *)obj);
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```
    return (int)__ldar32((unsigned __int32 volatile *)obj);
```

File: Include/cpython/pyatomic_std.h [hunk 5]
```diff
@@ -854,6 +854,23 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
+    _Py_USING_STD;
+    atomic_store_explicit((_Atomic(int)*)obj, value,
+                          memory_order_release);
+}
+
+static inline int
+_Py_atomic_load_int_acquire(const int *obj)
+{
+    _Py_USING_STD;
+    return atomic_load_explicit((const _Atomic(int*)*)obj,
```
Comment (colesbury) [inline Include/cpython/pyatomic_std.h]: I think this should be: `const _Atomic(int)*`

File: Python/ceval_gil.c [hunk 6]
```diff
@@ -205,8 +200,8 @@ static void destroy_gil(struct _gil_runtime_state *gil)
     COND_FINI(gil->switch_cond);
     MUTEX_FINI(gil->switch_mutex);
 #endif
-    _Py_atomic_store_explicit(&gil->locked, -1,
-                              _Py_memory_order_release);
+    _Py_atomic_store_int_relaxed(&gil->locked, -1);
+    _Py_atomic_store_int_release(&gil->locked, -1);
```
Comment (colesbury) [inline Python/ceval_gil.c:204]: Should only have the `_Py_atomic_store_int_release` call

**Review by colesbury — COMMENTED** [review]
Can you also add a test to `Modules/_testcapi/pyatomic.c` for the new functions?

File: Include/cpython/pyatomic_msc.h [hunk 7]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
 
+static inline void
+_Py_atomic_store_int_release(int *obj, int value)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    *(int volatile *)obj = value;
+#elif defined(_M_ARM64)
+    __stlr32((unsigned __int32 volatile *)obj, (unsigned __int32)value);
+#else
+#  error "no implementation of _Py_atomic_store_ptr_release"
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]: Error message should say "_Py_atomic_store_int_release"

File: Include/cpython/pyatomic_msc.h [hunk 8]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
+
+static inline int
+_Py_atomic_load_int_acquire(const int *obj)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    return *(int volatile *)obj;
+#elif defined(_M_ARM64)
+    return (int)__ldar32((unsigned __int32 volatile *)obj);
+#else
+#  error "no implementation of _Py_atomic_load_ptr_acquire"
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]: Error message should say "_Py_atomic_load_int_acquire"

File: Python/ceval_gil.c (hunk 6 above)
Comment (corona10) [reply Python/ceval_gil.c:204]: Thanks, this is my mistake.

**Review by colesbury — COMMENTED** [review]
Two more comments I missed in the first pass. When you get everything working to your satisfaction, I'd suggest testing the PR on all the buildbots as there is some platforms specific code.

**Comment by corona10** [comment]
> Two more comments I missed in the first pass. When you get everything working to your satisfaction, I'd suggest testing the PR on all the buildbots as there is some platforms specific code.

Got it

**Comment by corona10** [comment]
```
Python/ceval_gil.c: In function ‘create_gil’:
Python/ceval_gil.c:188:5: error: implicit declaration of function ‘_Py_ANNOTATE_RWLOCK_CREATE’ [-Werror=implicit-function-declaration]
     _Py_ANNOTATE_RWLOCK_CREATE(&gil->locked);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~
Python/ceval_gil.c: In function ‘destroy_gil’:
Python/ceval_gil.c:204:5: error: implicit declaration of function ‘_Py_ANNOTATE_RWLOCK_DESTROY’ [-Werror=implicit-function-declaration]
     _Py_ANNOTATE_RWLOCK_DESTROY(&gil->locked);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
Python/ceval_gil.c: In function ‘drop_gil’:
Python/ceval_gil.c:243:5: error: implicit declaration of function ‘_Py_ANNOTATE_RWLOCK_RELEASED’ [-Werror=implicit-function-declaration]
     _Py_ANNOTATE_RWLOCK_RELEASED(&gil->locked, /*is_write=*/1);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
Python/ceval_gil.c: In function ‘take_gil’:
Python/ceval_gil.c:356:5: error: implicit declaration of function ‘_Py_ANNOTATE_RWLOCK_ACQUIRED’ [-Werror=implicit-function-declaration]
     _Py_ANNOTATE_RWLOCK_ACQUIRED(&gil->locked, /*is_write=*/1);
```

Okay we catch one.

**Comment by vstinner** [comment]
Good idea to test on buildbots.

**Comment by corona10** [comment]
Convert to the draft PR since I need to check my things.

**Comment by vstinner** [comment]
You can use ``!buildbot PATTERN`` to only run tests on a specific CI rather pulling the huge number of CI, it may be faster as well. Tell me of you need help about that.

**Comment by corona10** [comment]
@colesbury @vstinner 
Okay, failed test(test_exception) on build bots are not directly related to this PR. (I monitored another build, and it happens)

File: Include/cpython/pyatomic_msc.h [hunk 9]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
 #endif
 }
 
+static inline void
+_Py_atomic_store_int_release(int *obj, int value)
+{
+#if defined(_M_X64) || defined(_M_IX86)
+    *(int volatile *)obj = value;
```
Comment (corona10) [inline Include/cpython/pyatomic_msc.h:919]: @colesbury 
Basically, int volatile * and volatile int* have the same meaning.
So I just follow the convention from `_Py_atomic_load_ptr_acquire`

File: Modules/_testcapi/pyatomic.c [hunk 10]
```diff
@@ -140,6 +140,22 @@ test_atomic_release_acquire(PyObject *self, PyObject *obj) {
     Py_RETURN_NONE;
 }
 
+static PyObject *
+test_atomic_load_store_int_release_acquire(PyObject *self, PyObject *obj) { \
+    int x = 0;
+    int y = 1;
+    int z = 2;
+    _Py_atomic_load_int_acquire(&x);
+    assert(x == 0);
```
Comment (vstinner) [inline Modules/_testcapi/pyatomic.c]: This test is strange. You should test the load() result, no?

File: Python/ceval_gil.c [hunk 11]
```diff
@@ -1,6 +1,6 @@
 
 #include "Python.h"
-#include "pycore_atomic.h"        // _Py_atomic_int
+#include "pycore_atomic.h"        // _Py_ANNOTATE_RWLOCK_XXXX
```
Comment (vstinner) [inline Python/ceval_gil.c]: I would prefer to mention a real symbol:

```suggestion
#include "pycore_atomic.h"        // _Py_ANNOTATE_RWLOCK_CREATE
```

Same remark in Python/thread_pthread.h.

File: Include/cpython/pyatomic_msc.h [hunk 12]
```diff
@@ -912,6 +912,30 @@ _Py_atomic_store_ptr_release(void *obj, void *value)
...
+#elif defined(_M_ARM64)
+    __stlr32((unsigned __int32 volatile *)obj, (unsigned __int32)value);
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h:922]: It terms of memory alignment, is it safe to cast `int*` to `int32*`? Is it the same size?

Please add:

```
_Py_atomic_ASSERT_ARG_TYPE(unsigned __int32);
```

**Review by vstinner — APPROVED** [review]
LGTM

**Review by colesbury — APPROVED** [review]

**Pull request merged**